### PR TITLE
feat: GRD-70189 Supporting SQS input in AWS Dynamo DB filter plugin.

### PR DESCRIPTION
**Pull request by chirag-ibm**
[GRD-70189](https://jira-ds.rtp.raleigh.ibm.com:8443/browse/GRD-70189)
This PR includes the code changes required to support the logs pull of AWS Dynamo DB through the AWS SQS. In summary it covers,

Modifications in the backend java code and hence in the Zip package.
New configurations example which uses Logstash SQS input.
New points in added in the README documentation that explains the procedures of setting up Lambda functions and creating the SQS.

### CODE REVIEW

**Review by PratikshaSonawane — APPROVED** [review]

**Review by ofer-haim — APPROVED** [review]
Please do another quick review to verify it's the exact same change as verified by QA before merge

File: filter-plugin/logstash-filter-dynamodb-guardium/CHANGELOG.md [hunk 1]
```diff
@@ -4,9 +4,14 @@ Notable changes will be documented in this file.
 ## [Unreleased]
 
 ## [1.0.0] - 2021-05-17
-
 ### Added
 - Initial release, in parallel to Guardium v11.3.
 
+## [1.0.1] - 2024-02-01
+### Added
+- Support of SQS input. The plugin will be able to read and process the Dynamo DB
```
Comment (mllezak) [inline filter-plugin/logstash-filter-dynamodb-guardium/CHANGELOG.md:12]: Support of Simple Queue Service (SQS) input. This plugin can read and process the DynamoDB

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md [hunk 2]
```diff
@@ -3,20 +3,30 @@
...
-      * Supported inputs:
-        * CloudWatch (pull)
-  * Guardium Insights SaaS: 1.0
-      * Supported inputs:
-        * CloudWatch (pull)
-
-This is a [Logstash](https://github.com/elastic/logstash) filter plug-in for the universal connector that is featured in IBM Security Guardium. It parses events and messages from the Amazon DynamoDB audit log into a [Guardium record](https://github.com/IBM/universal-connectors/blob/main/common/src/main/java/com/ibm/guardium/universalconnector/commons/structures/Record.java) instance (which is a standard structure made out of several parts). The information is then sent over to Guardium. Guardium records include the accessor (the person who tried to access the data), the session, data, and exceptions. If there are no errors, the data contains details about the query "construct". The construct details the main action (verb) and collections (objects) involved.
-
-The plug-in is free and open-source (Apache 2.0). It can be used as a starting point to develop additional filter plug-ins for Guardium universal connector.
+    * Guardium Data Protection: 11.4 and above
```
Comment (mllezak) [inline filter-plugin/logstash-filter-dynamodb-guardium/README.md:6]: 11.4 and later.

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md [hunk 3]
```diff
@@ -70,30 +80,149 @@ There are different ways to audit and log events. We will use CloudTrail for thi
...
-#### Note
-Before making any changes to the lambda function code, first disable the rule discussed previously.
-Deploy the change and then re-enable the rule.
+## 4. Exporting Cloudwatch Logs to SQS using lambda function
+In order to achieve load balancing of audit logs between different collectors,
+the audit logs must be exported from Cloudwatch to SQS.
 
-### Limitations
+### Creating the SQS queue
+The SQS created in these steps will contain the messages to be filled up by the
```
Comment (mllezak) [inline filter-plugin/logstash-filter-dynamodb-guardium/README.md:95]: I don't quite understand -- is this clearer?

The SQS that you create contains messages that will be filled by the Lambda function (see [Creating the Lambda function](Creating the Lambda function)) by reading the CloudWatch logs.

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md [hunk 4]
```diff
@@ -70,30 +80,149 @@ There are different ways to audit and log events. We will use CloudTrail for thi
...
+    4. Click **Attach existing policies directly**.
+    5. Search for the policy created and check the checkbox next to it.
+    6. Click **Next: Review**
+    7. Click **Add permissions**
+
+### Creating the Lambda function
+The Lambda function will read the CloudWatch Logs and send the events into the SQS queue.
+Follow the steps below to configure the Lambda function.
+
+#### Creating IAM Role
```
Comment (mllezak) [inline filter-plugin/logstash-filter-dynamodb-guardium/README.md:140]: Create the IAM Role

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md [hunk 5]
```diff
@@ -70,30 +80,149 @@ There are different ways to audit and log events. We will use CloudTrail for thi
...
+2. Go to **IAM** -> **Roles**
+3. Click **Create Role**
+4. Under use case select **Lambda** and click **Next**
+5. Search for **AmazonSQSFullAccess** and select it
+6. Search for **CloudWatchLogsFullAccess** and select it
+7. Search for **CloudWatchEventsFullAccess** and select it
+8. Set the Role Name: e.g., "Export-Dynamo-CloudWatch-to-SQS-Lambda" and click **Create role**.
+
+
+#### Create the lambda function
```
Comment (mllezak) [inline filter-plugin/logstash-filter-dynamodb-guardium/README.md:156]: Create the Lambda function

(Note -- I think AWS Lambda is capitalized -- we just need to be consistent).

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md [hunk 6]
```diff
@@ -70,30 +80,149 @@ There are different ways to audit and log events. We will use CloudTrail for thi
...
+4. Click **Create Queue**.
+5. Select the type as **Standard**.
+6. Enter the name for the queue
+7. Keep the rest of the default settings
+
+### Creating a policy for the relevant IAM User
+Perform the following steps for the IAM user who is accessing the SQS logs in
+Guardium:
+#### Procedure
+1. Go to https://console.aws.amazon.com/
```
Comment (mllezak) [inline filter-plugin/logstash-filter-dynamodb-guardium/README.md:111]: Suggestion for all of the procedures where the first steps is Go to https://console.aws.amazon.com/...

Fold step 1 and step 2 into a single step, for example: 

1. From https://console.aws.amazon.com/, click **Services**.

Or 

1. From https://console.aws.amazon.com/, browse to **IAM** -> **Roles**.

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md [hunk 7]
```diff
@@ -70,30 +80,149 @@ There are different ways to audit and log events. We will use CloudTrail for thi
...
+9. Click **Create function** and navigate to **Code view**.
+10. Add the function code from the file [lambda.py](DynamodbOverSQSPackage/lambda.py).
+11. Click **Configuration** > **Environment Variables**.
+12. Create 2 variables:
+    1. `Key = GROUP_NAME value = <value>` e.g., `uc-dynamo-plugin-cloudwatch-group`
+    2. `Key = QUEUE_NAME value = <value>` e.g., `https://sqs.ap-south-1.amazonaws.com/346824951129/np-dynamo-sqs`
+13. Save the function.
+14. Click **Deploy**.
+
+#### Automating the lambda function
```
Comment (mllezak) [inline filter-plugin/logstash-filter-dynamodb-guardium/README.md:176]: #### Automate the Lambda function

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md [hunk 8]
```diff
@@ -70,30 +80,149 @@ There are different ways to audit and log events. We will use CloudTrail for thi
...
+The Lambda will be called by a scheduler configured inside event rules in CloudWatch.
+
+**_Procedure_**
+1. Go to the CloudWatch dashboard.
```
Comment (mllezak) [inline filter-plugin/logstash-filter-dynamodb-guardium/README.md:180]: From the CloudWatch dashboard, browse to **Events** > **Rules**.

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md [hunk 9]
```diff
@@ -70,30 +80,149 @@ There are different ways to audit and log events. We will use CloudTrail for thi
...
+2. Go to **Events** > **Rules**.
+3. Click **Create Rule**.
+4. Enter the name for the rule e.g., cloudwatchToSqs
+5. Under **Rule Type**, select **Schedule**.
+6. Define the schedule. In **schedule pattern**, select a schedule that runs at a regular rate,
+   such as every 10 minutes.
+7. Enter the rate expression, meaning the rate at which the function should execute.
+   This value must match the time specified in the lambda function code that calculates the time delta.
```
Comment (mllezak) [inline filter-plugin/logstash-filter-dynamodb-guardium/README.md:188]: **Note:** This value must match the time that is specified in the Lambda function code that calculates the time delta. (For example, if the function code is set to 2 minutes, set the rate to 2 minutes.)

8. Click **Next**.

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md [hunk 10]
```diff
@@ -70,30 +80,149 @@ There are different ways to audit and log events. We will use CloudTrail for thi
...
+10. Select the lambda function created in the above step. e.g., Export-Dynamo-CloudWatch-Logs-To-SQS.
+11. Add the tag if needed.
+12. Click **Create Rule**.
 
-1. The DynamoDB plug-in does not support IPV6.
+#### Note
+* Before making any changes to the lambda function code, first disable the above rule.
+  Deploy the change and then re-enable the rule.
+* If the **Management events** are enabled while configuring the **CloudTrail**, it is possible
+  that the data being read by the Lambda Function would be large. In such cases the `timeout` parameter
```
Comment (mllezak) [inline filter-plugin/logstash-filter-dynamodb-guardium/README.md:200]: that a large amount of data will be read by the Lambda function. In this case you might need to tune the `timeout` and `memory` parameters.

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md [hunk 11]
```diff
@@ -70,30 +80,149 @@ There are different ways to audit and log events. We will use CloudTrail for thi
...
+    * Go to **General Configurations**.
+    * Click **Edit**.
+    * Update the value of **Memory** as required.
+    * Update the value of **Timeout** as required.
 
-2. You may need to disable management events in order to avoid heavy traffic and data loss in Guardium. Disabling management events disables logging of the following events: 
-CreateTable, DeleteTable, ListTable, UpdateTable, DescribeTable events.
+### Limitations
 
+1. The Dynamo DB plug-in does not support IPV6.
```
Comment (mllezak) [inline filter-plugin/logstash-filter-dynamodb-guardium/README.md:211]: DynamoDB -- always one word.

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md [hunk 12]
```diff
@@ -112,36 +241,39 @@ grdapi add_domain_to_universal_connector_allowed_domains domain=amazon.com
 
 • Download the [cloudtrail_codec_plugin.zip](DynamodbOverCloudtrailPackage/DynamoDB/cloudtrail_codec_plugin.zip) plug-in, if the logs are to be pulled directly from CloudTrail. (Do not unzip the offline-package file throughout the procedure). This step is not necessary for Guardium Data Protection v12.0 and later.
 
-• Download the [logstash-filter-dynamodb_guardium_plugin_filter.zip plug-in.](https://github.com/IBM/universal-connectors/releases/download/v1.5.1/logstash-filter-dynamodb_guardium_plugin_filter.zip) (Do not unzip the offline-package file throughout the procedure).
+• Download the [logstash-filter-dynamodb_guardium_plugin_filter.zip plug-in.](./logstash-filter-dynamodb_guardium_plugin_filter.zip)
 
 ### Procedure
 1. On the collector, go to **Setup** > **Tools and Views** > **Configure Universal Connector**.
-2. Enable the connector if it is disabled.
+2. Enable the connector if it is already disabled, before proceeding to upload the UC.
```
Comment (mllezak) [inline filter-plugin/logstash-filter-dynamodb-guardium/README.md:248]: Before uploading the Universal Connector, make sure that it is enabled.

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md [hunk 13]
```diff
@@ -112,36 +241,39 @@ grdapi add_domain_to_universal_connector_allowed_domains domain=amazon.com
...
-		1.  [cloudtrail_codec_plugin.zip](./DynamodbOverCloudtrailPackage/DynamoDB/cloudtrail_codec_plugin.zip)
-      plug-in. After it is uploaded, click **OK**. This step is not necessary for Guardium Data Protection v12.0 and later.
-		2.  [logstash-filter-dynamodb_guardium_plugin_filter.zip](https://github.com/IBM/universal-connectors/releases/download/v1.5.1/logstash-filter-dynamodb_guardium_plugin_filter.zip)
-      plug-in. After it is uploaded, click **OK**. 
-      
+    * If the audit logs are to be fetched from CloudWatch directly, 
+      select the [logstash-filter-dynamodb_guardium_plugin_filter.zip](https://github.com/IBM/universal-connectors/releases/download/v1.5.1/logstash-filter-dynamodb_guardium_plugin_filter.zip) plug-in. After it is uploaded, click **OK**.
+    * If the audit logs are to be fetched from CloudTrail, select the
+      1. [cloudtrail_codec_plugin.zip](./DynamodbOverCloudtrailPackage/DynamoDB/cloudtrail_codec_plugin.zip) plug-in. After it is uploaded, click **OK**. 
+       This step is not necessary for Guardium Data Protection v12.0 and later.
```
Comment (mllezak) [inline filter-plugin/logstash-filter-dynamodb-guardium/README.md:254]: **Note:** This step is not necessary for Guardium Data Protection 12.0 or later.

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md [hunk 14]
```diff
@@ -112,36 +241,39 @@ grdapi add_domain_to_universal_connector_allowed_domains domain=amazon.com
...
-6. Update the input section to add the details from the corresponding file's input part, omitting the keyword "input{" at the beginning and its corresponding "}" 
-	at the end.
-	1. If the audit logs are to be fetched from CloudWatch, use the details from the [dynamodbCloudwatch.conf](./dynamodbCloudwatch.conf)
-    	file. More details on how to configure the relevant input plug-in can be found [here](../../input-plugin/logstash-input-cloudwatch-logs/README.md).
-    2. If the audit logs are to be fetched from CloudTrail, use the details from the [dynamodbCloudtrail.conf](./dynamodbCloudtrail.conf) 
-    	file. 
+6. Update the input section to add the details from the corresponding file's input part, omitting the keyword "input{" at the beginning and its corresponding "}"
+   at the end.
+    1. If the audit logs are to be fetched from CloudWatch, use the details from the [dynamodbCloudwatch.conf](./dynamodbCloudwatch.conf)
+       file. More details on how to configure the relevant input plug-in can be found [here](../../input-plugin/logstash-input-cloudwatch-logs/README.md).
```
Comment (mllezak) [inline filter-plugin/logstash-filter-dynamodb-guardium/README.md:262]: For more details about configuring the relevent input plugin, see ....

**Review by mllezak — CHANGES_REQUESTED** [review]
Some general comments:

Make sure  that you use consistent capitalization and spelling: 

* Lambda function
* DynamoDB

Line 95, I didn't understand what you meant -- I tried to rewrite it, but I might not understand enough. 

Line 111, see my suggestion about folding steps 1 and 2 together -- this is a very long README and there are a lot of steps -- if we can remove one step for each procedure, it's just a little easier to read.

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md [hunk 15]
```diff
@@ -35,7 +45,7 @@ In the AWS web interface, configure the service for DynamoDB.
 8. Enter a partition key.
 9. Scroll down and click **Create table**.
 
-## 2. Enabling audit logs 
+## 2. Enabling audit logs
 
 There are different ways to audit and log events. We will use CloudTrail for this example since it supports all required parameters. The following events are supported for auditing in AWS:
```
[the remark this answers is not in the document]
Comment (chirag-ibm) [reply filter-plugin/logstash-filter-dynamodb-guardium/README.md:50]: I think this text is getting carried forward, Right now we almost support all the Events which are falling under Dynamo DB specific events. I will remove this line.

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md (hunk 3 above)
Comment (chirag-ibm) [reply filter-plugin/logstash-filter-dynamodb-guardium/README.md:95]: Agreed, The new content is clearer.

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md [hunk 16]
```diff
@@ -70,30 +80,149 @@ There are different ways to audit and log events. We will use CloudTrail for thi
...
+8. Select **Target1**. Select the **Target Type** as **AWS Service**.
+9. Select **Target** as **Lambda Function**.
+10. Select the Lambda function created in the above step. e.g., Export-Dynamo-CloudWatch-Logs-To-SQS.
+11. Add the tag if needed.
+12. Click **Create Rule**.
+
 #### Note
-Before making any changes to the lambda function code, first disable the rule discussed previously.
-Deploy the change and then re-enable the rule.
+* Before making any changes to the Lambda function code, first disable the above rule.
```
Comment (mllezak) [inline filter-plugin/logstash-filter-dynamodb-guardium/README.md:197]: Before you make changes to the Lambda function code, disable the rule that you just created.

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md [hunk 17]
```diff
@@ -70,30 +80,149 @@ There are different ways to audit and log events. We will use CloudTrail for thi
...
 ### Limitations
 
 1. The DynamoDB plug-in does not support IPV6.
-
-2. You may need to disable management events in order to avoid heavy traffic and data loss in Guardium. Disabling management events disables logging of the following events: 
-CreateTable, DeleteTable, ListTable, UpdateTable, DescribeTable events.
-
+2. You may need to disable management events in order to avoid heavy traffic and data loss in Guardium. Disabling management events disables logging of the following events:
+   CreateTable, DeleteTable, ListTable, UpdateTable, DescribeTable events.
+3. The following fields couldn't be mapped with the Dynamo audit logs,
```
Comment (mllezak) [inline filter-plugin/logstash-filter-dynamodb-guardium/README.md:214]: The `Client HostName` field cannot be mapped with the DynamoDB audit logs, Set this field to *NA*.

**Review by mllezak — APPROVED** [review]
I had just a few more comments. I approve after those are made.

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md [hunk 18]
```diff
@@ -47,8 +57,8 @@ There are different ways to audit and log events. We will use CloudTrail for thi
 4. Click **Create trail** button.
 5. Provide a trail name in the **Trail name** field.
 6. For **Storage location**, verify that **Create new S3 bucket** is selected.
-	 The logs can be read directly from this bucket using the S3 input of the logstash.
-7. In **Log file SSE-KMS encryption**, uncheck **Enabled**. 
+   The logs can be read directly from this bucket using the S3 input of the logstash.
+7. In **Log file SSE-KMS encryption**, uncheck **Enabled**.
 8. If the logs are to be monitored through CloudWatch, then forward them to CloudWatch using steps 9 to 13. (If not, skip those steps).
```
Comment (mllezak) [inline filter-plugin/logstash-filter-dynamodb-guardium/README.md:62]: 8. To monitor the logs through CloudWatch, take the following steps to forward them to Cloudwatch:
   1. Under **CloudWatch Logs**, check **Enabled**.
   2. Verify that **New** is selected for **Log group**.
   3. Under **Log group name**, provide a new log group name.
   4. Verify that **New** is selected for **IAM Role**.
   5. For **Role name**, provide a new role name.

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md [hunk 19]
```diff
@@ -70,30 +80,148 @@ There are different ways to audit and log events. We will use CloudTrail for thi
...
+### Creating the SQS queue
+The SQS that you create contains messages that will be filled by the Lambda 
+function (see [Creating the Lambda function](#creating-the-lambda-function)) by reading the 
+CloudWatch logs.
+#### Procedure
+1. From https://console.aws.amazon.com/, click **Services**.
+2. Search for SQS and click on **Simple Queue Services**
+3. Click **Create Queue**.
+4. Select the type as **Standard**.
+5. Enter the name for the queue
```
Comment (mllezak) [inline filter-plugin/logstash-filter-dynamodb-guardium/README.md:103]: Add a period after steps 5 and 6.

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md [hunk 20]
```diff
@@ -70,30 +80,148 @@ There are different ways to audit and log events. We will use CloudTrail for thi
...
+The Lambda function will read the CloudWatch Logs and send the events into the SQS queue.
+Follow the steps below to configure the Lambda function.
+
+#### Create IAM Role
+Create the IAM role that will be used in the Lambda function setup. The AWS Lambda service requires permission to log 
+events and write to the SQS queue. Create the IAM Role **Export-Dynamo-CloudWatch-to-SQS-Lambda** with
+"AmazonSQSFullAccess", "CloudWatchLogsFullAccess", and "CloudWatchEventsFullAccess" policies.
+
+__*Procedure*__
+1. Go to https://console.aws.amazon.com/
```
Comment (mllezak) [inline filter-plugin/logstash-filter-dynamodb-guardium/README.md:145]: 1. From https://console.aws.amazon.com/, browse to  **IAM** -> **Roles**.

File: filter-plugin/logstash-filter-dynamodb-guardium/README.md [hunk 21]
```diff
@@ -70,30 +80,148 @@ There are different ways to audit and log events. We will use CloudTrail for thi
...
+2. Go to **IAM** -> **Roles**
```
Comment (mllezak) [inline filter-plugin/logstash-filter-dynamodb-guardium/README.md:146]: delete step 2.

**Review by mllezak — APPROVED** [review]
Nothing terrible here -- just some more cleanup (and trying to shorten the number of steps in some of the longer procedures).

Approved with changes.

**Pull request merged**